Throwable toString override

daddykotex opened the pull request:
## PR Checklist (not all items are relevant to all PRs)

- [ ] Added unit-tests (for runtime code)
- [x] Update bootstrapped code + smoke tests (when the rendering logic is modified)
- [ ] Added build-plugins integration tests (when reflection loading is required at codegen-time)
- [ ] Added alloy compliance tests (when simpleRestJson protocol behaviour is expanded/updated)
- [ ] Updated dynamic module to match generated-code behaviour
- [ ] Added documentation
- [x] Updated changelog


0.17 backport of https://github.com/disneystreaming/smithy4s/pull/1263

daddykotex committed:
Override the default toString when extends Throwable

daddykotex committed:
Update examples

kubukoz commented on modules/codegen/src/smithy4s/codegen/internals/LineSegment.scala:
```diff
@@ -73,6 +73,13 @@ private[codegen] object LineSegment {
+      "scala",
+      "java.lang",
+      "scala.Predef",
+      "scala.collection.immutable"
```
**question:** should this be here? I don't think scala imports this by default, correct me if I'm wrong but some things like `List`/`Set`/`Map` are available because of aliases in `scala` / `scala.Predef`.

daddykotex replied:
You're probably right, I grabbed that code from the 0.18 branch.

If I remove it, some files in `bootstrapped` will get an additional import for things you mentionned: like `import scala.collection.immutable.Map`. So although it might be incorrect, the incorrect behaviour does not show because we don't use shapes that would be in `scala.collection.immutable` but not imported by the predef.

kubukoz replied:
Can we change occurrences of `NameRef("scala.collection.immutable", "Map")` to `NameRef("Predef", "Map")`? wouldn't that give us the right behavior too?

kubukoz replied:
btw. this sent me on a spiral of madness because I remembered some 2.12-specific issue with immutable/general collections. Thought it was about Set/Map, but it was about `Seq`:

from [2.13 release notes](https://github.com/scala/scala/releases/tag/v2.13.0):

```
Immutable scala.Seq
    Seq is now an alias for collection.immutable.Seq
        Before, it was an alias for the possibly-mutable collection.Seq.
```

kubukoz commented on modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:
```diff
@@ -663,6 +665,9 @@ private[internals] class Renderer(compilationUnit: CompilationUnit) { self =>
   }
 
+  private def defaultThrowableGetMessage =
+    line"override def toString(): $string_ = $scalaRuntime._toString(this)"
```
**issue:** needs a changelog entry (on both versions) ;)

kubukoz commented on modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:
```diff
@@ -663,6 +665,9 @@ private[internals] class Renderer(compilationUnit: CompilationUnit) { self =>
     )
   }
 
+  private def defaultThrowableGetMessage =
```
**suggestion:** I would add a code comment with your explanation from the 0.18 PR:

> Default `toString` on `Throwable` prints the classname. This is suboptimal when we could have the case class `toString` implementation available.

daddykotex replied:
done, good suggestion

daddykotex committed:
Add changelog entry

Baccata commented on modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:
```diff
@@ -663,6 +665,10 @@ private[internals] class Renderer(compilationUnit: CompilationUnit) { self =>
 
+  // Default toString on Throwable prints the classname, instead we use the case class toString.
+  private def defaultThrowableGetMessage =
+    line"override def toString(): $string_ = $scalaRuntime._toString(this)"
```
See my comment [there](https://github.com/disneystreaming/smithy4s/pull/1263/files#r1377405904)

daddykotex committed:
Revert "Override the default toString when extends Throwable"

This reverts commit 396683d5fd87da9297f18ef3436a1dafbeb4c8af.

daddykotex committed:
Render custom toString

I chose to implement toString, because implementing getMessage would
lead to `smithy4s.example.ClientError: smithy4s.example.ClientError(400, "oops")`
which felt weird

Baccata commented:
Closing this one : we're not promising LTS in any way.
